File: leetcode/SymmetricTree/isSymmetric_recursion.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isSymmetric(TreeNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if (root == null) return true;
        return helper(root.left, root.right);
    }
    private boolean helper(TreeNode a, TreeNode b) {
        if (a == null || b == null) return a == b;
        return a.val == b.val && helper(a.left, b.right) && helper(a.right, b.left);
    }
}

File: leetcode/SymmetricTree/isSymmetric_recursion.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return !root || helper(root->left, root->right);
    }
    bool helper(TreeNode *left, TreeNode *right) {
        if (!left || !right) return left == right;
        return (left->val == right->val) && helper(left->left, right->right) && helper(left->right, right->left);
    }
};
